src/server/pegasus_server_impl.cpp

```diff
@@ -856,7 +900,10 @@ void pegasus_server_impl::on_multi_get(const ::dsn::apps::multi_get_request &req
                     }
                 }
 
-                iterate_count++;
+                limiter->add_count();
+                if (!limiter->time_check()) {
```

Comment:
You can include `time_check()` in `valid()`.

I guess why you separate the two calls is that `time_check` depends on `add_count`. Because in the first iteration, the condition always meets `_iteration_count % _module_num == 0`, so it checks time duration futilely.

If you want every "module_num" times, the `dsn_now` actually calls, it needs a slight change:

```cpp
    // during rocksdb iteration, if iteration_count % module_num == 0, we will check if iteration
    // exceed time threshold, which means we at most check ten times during iteration
    bool time_check()
    {
        if (_max_duration_time > 0 &&  (_iteration_count+1) % _module_num == 0 &&
            dsn_now_ns() - _iteration_start_time_ns > _max_duration_time) {
            _exceed_limit = true;
            _iteration_duration_time_ns = dsn_now_ns() - _iteration_start_time_ns;
            return false;
        }
        return true;
    }
```

So that only after (module_num-1) will there be the first call of dsn_now.

```cpp
    bool valid()
    {
        if (_iteration_count >= _max_count) {
            return false;
        }
        if (_max_size > 0 && _iteration_size >= _max_size) {
            return false;
        }
        return time_check();
    }
```

Reply:
Your conjecture is correct and just is part of the reason that I used to separate adding iteration count. 
Iteration_count will not always be auto-increment during iteration.

Reply:
I misunderstand your suggestion before, the code is updated.

Reply:
I misunderstand your suggestion before, the code is updated.